### ISSUE TITLE
fail to train joint_alignment_translation model with fp16

### ISSUE THREAD

**Issue by YonghyunRyu** [facebookresearch/fairseq#4269]
## 🐛 Bug

joint_alignment_translation example causes a gradient explosion error when training it with fp16 option.
(https://github.com/pytorch/fairseq/tree/main/examples/joint_alignment_translation)

### To Reproduce
I use 8 A100s or 8 V100s for training with cudatoolkit 11.3.
I tried both older and latest versions of fairseq. torch 1.8.2 and latest, tried both python 3.8 and 3.9.

#### Code sample
I only followed the given instructions in the README.md

### Expected behavior
training to converge

### Environment
 - fairseq Version 1.0 and 0.10.2
 - PyTorch Version 1.8.2 and 1.10.2
 - OS (e.g., Linux): Ubuntu 20.04
 - How you installed fairseq (`pip`, source): install from source and pip repos
 - Build command you used (if compiling from source):
 - Python version: 3.8.12 and 3.9.7
 - CUDA/cuDNN version: 11.3.1
 - GPU models and configuration: A100 or V100
 - Any other relevant information:

### Additional context
I tried large batch, grad clipping, scaling learning rate and so on, but all failed.
But without `--fp16`, it works

**Comment by melody-rain** [comment]
This seems to be an old issue, which I have not seen any solutions.
